src/modules/simulator/simulator_mavlink.cpp

```diff
@@ -913,6 +928,116 @@ int openUart(const char *uart_name, int baud)
 }
 #endif
 
+void Simulator::check_failure_injections()
+{
+	vehicle_command_s vehicle_command;
+
+	while (_vehicle_command_sub.update(&vehicle_command)) {
+		if (vehicle_command.command != vehicle_command_s::VEHICLE_CMD_INJECT_FAILURE) {
+			continue;
+		}
+
+		bool handled = false;
+		bool supported = false;
+
+		const int failure_unit = static_cast<int>(vehicle_command.param1 + 0.5f);
+		const int failure_type = static_cast<int>(vehicle_command.param2 + 0.5f);
+
+		if (failure_unit == vehicle_command_s::FAILURE_UNIT_SENSOR_GPS) {
```

Comment:
These nested ifs get a little nasty. It would be nice to have a separation between the message handling logic, and the definitions of which failure unit supports what mode. Maybe have a static array for the failure units with a bitmask for what mode each one supports, similar to the names you did below? Then just check that the requested action is supported, a switch to get a `bool*` to the correct failure unit, and a second switch to apply the correct update to that `bool *`.

This is more a style thing though, not critical.

Reply:
True that would be nicer! This logic is likely going to be removed/rewritten and moved to the sensors module, so I'm reluctant to change it now.